lib/puppet/state_machine.rb

```diff
@@ -0,0 +1,197 @@
+module Puppet
+  # Define state machines for modeling complex operations.
+  #
+  # A given state machine is immutable once built; invoking the state machine generates a context
+  # for that invocation that leaves the state machine and associated states unmodified.
+  class StateMachine
+    require 'puppet/state_machine/state'
+    require 'puppet/state_machine/machine_builder'
+    require 'puppet/state_machine/machine_runner'
+
+    # Define, verify, and build a new state machine
+    #
+    # @example
+    #   Puppet::StateMachine.build("Example Puppet runner") do |m|
+    #     m.start_state(:start)
+    #
+    #     m.state(:start,
+    #       action: -> { Puppet.notice("Starting Puppet agent") },
+    #       event: ->(result) { :create_lockfile },
+    #       transitions: {create_lockfile: :create_lockfile})
+    #
+    #     m.state(:create_lockfile,
+    #       action: -> { Puppet::Util::Lockfile.new("/var/lock/puppet.lock").lock },
+    #       event: ->(has_lock) { has_lock ? :got_lock : :already_locked },
+    #       transitions: {
+    #         got_lock: :run_agent,
+    #         already_locked: :cancel_run,
+    #       })
+    #
+    #     m.state(:cancel_run,
+    #       type: :error,
+    #       action: -> { Puppet.error("Cancelling run.") }
+    #
+    #     m.state(:run_puppet,
```

Comment:
I think this state should be `:run_agent` given line 26? Really quite minor but I'm a stickler for doc strings being accurate.